source/tutorials/module-system/module-system.md

```diff
@@ -0,0 +1,1027 @@
+# Module system introduction
+
+Note: This tutorial was created from https://github.com/tweag/summer-of-nix-modules, as presented by @infinisil at Summer of Nix 2021, presentation can be seen here (might not work in browser, `mpv` should work): https://infinisil.com/modules.mp4
+
+## Empty module
+
+The simplest module you can have is just an empty attribute set, which
+as you might expect, doesn't do anything!
+
+```diff
+diff --git a/default.nix b/default.nix
+new file mode 100644
+index 0000000..1797133
+--- /dev/null
++++ b/default.nix
+@@ -0,0 +1,3 @@
++{
++  
++}
+```
+
+## Module arguments: lib
+
+Modules can be just an attribute set. But if you want access to some
+arguments you need to change it into a function taking an attribute set
+with an ellipsis (that's the "..."). In this case we only match on the
+`lib` argument, which gives us access to nixpkgs library functions.
+
+Note that the ellipsis is necessary since arbitrary arguments can be
+passed to modules.
+
+```diff
+diff --git a/default.nix b/default.nix
+index 1797133..f7569bd 100644
+--- a/default.nix
++++ b/default.nix
+@@ -1,3 +1,3 @@
+-{
++{ lib, ... }: {
+   
+ }
+```
+
+## Declaring generate.script option
+
+In order for modules to be useful, we need to have options, so let's
+declare one. Options are declared by defining an attribute with
+`lib.mkOption` under the `options` attribute. Here we're defining the
+option `generate.script`.
+
+While there are many attributes to customize options, the most
```

Comment:
Where I can find those other attributes?

Reply:
Afaik the source has the only complete listing: https://github.com/NixOS/nixpkgs/blob/fa1d4611cc2ce41d7b93c4a8e8db0efaaf286475/lib/options.nix#L66-L88